README.md
% Checkpassword SOAP service  
% Fabrice Le Coz  
% February 2015

This little site is used to control the login/password fill-up by the user on the IDS platform.
 
This site is requested through the IP given by the IDS platform ( VPN on test )

So if the ip change the service must mutch this ip, to generate the correct wsdl.

# Install

install the project in `/home/http/SOAPservice`

    git clone git@git.telecomsante.com:physiodom-hdim/SOAPservice.git /home/http/SOAPservice

Create a config file in `/home/http/` using the `config.json.sample`

    cp /home/http/SOAPservice/config.json.sample /home/http/SOAPservice.json

edit the file to set the correct IP, in dev mode the IP is the IP given by the VPN with the IDS platform.

copy the init.d file and register the service, it will be better if the service starts before the HHR-pro instance

    sudo cp /home/http/SOAPservice/install/SOAPservice /etc/init.d
    sudo update-rc.d SOAPservice defaults 40

CHANGELOG.md
% CHANGELOG SOAPservice  
% Fabrice le Coz <fabrice.lecoz@telecomsante.com>  
% March, 2015

__v0.0.17__

  - Fix etcd watchdog : at start get all HHRPro instances
  - sets the timeout to 1000ms to query all HHRPro instances 

__v0.0.16__

  - Fix etcd watchdog
  
__v0.0.15__

  - Use etcd to get all HHRPro instances
  
__v0.0.14__

  - First version
  
> __Nota :__ the version number reflect the version number of the HHR-Pro portal

server.js
var http = require("http"),
    soap = require("soap"),
    promise = require("rsvp").Promise,
    Logger = require("logger"),
    request = require("request"),
    swig  = require('swig');

var logger = new Logger("WSDL service");
var serverIP = "10.122.31.126";

var CheckPassword = {
	loginservice: {
		loginserviceSOAP: {
			CheckPassword: function (CheckPasswordRequest, cb) {
				var soapReq = CheckPasswordRequest.CheckPasswordRequest;
				
				if (!soapReq || !soapReq.Authentifier || !soapReq.OrganizationalUnit || !soapReq.Password) {
					logger.error("Requete d'identification invalide");
					throw {
						Fault: {
							Code  : {
								Value  : "soap:Sender",
								Subcode: {value: "rpc:BadArguments"}
							},
							Reason: {Text: "Requete invalide"}
						}
					};
				}

				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};
				
				function testPasswd( Authentifier, Password ) {
					return new promise( function(resolve, reject) {
						request({
							url    : "http://192.168.1.68/api/checkpasswd",
							method : "POST",
							headers: {"content-type": "text/plain"},
							body   : JSON.stringify({login: Authentifier, password: Password})
						}, function (err, resp, body) {
							if (err) {
								logger.debug("resp checkpasswd err", err);
								resolve( { valid: false });
							} else {
								try {
									var isvalid = JSON.parse(body);
									resolve( isvalid );
								} catch( err ) {
									resolve( { valid: false });
								}
							}
						});
					});
				}
				
				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};
				// @todo valid is a function that validate the user:
				
				testPasswd( soapReq.Authentifier, soapReq.Password )
					.then( function( isvalid ) {
						var valid = isvalid.valid;
						if (valid) {
							logger.info("Authentifier valid", soapReq);
							CheckPasswordOut.IsValid = true;
							CheckPasswordOut.CheckPasswordUserInfo = "Authentifier=" + soapReq.Authentifier + ";OrganizationalUnit=" + soapReq.OrganizationalUnit + ";";
						} else {
							logger.alert("Authentifier not valid", soapReq);
							CheckPasswordOut.IsValid = false;
						}
						logger.trace( { CheckPasswordResponse: CheckPasswordOut } );
						cb({ CheckPasswordResponse: CheckPasswordOut });
					});
			}
		}
	}
};

var test = {};

test.test1 = function() {
	logger.trace("test1 : get a soap client");
	return new promise( function( resolve, reject) {
		soap.createClient("http://"+serverIP+'/wsdl', function(err, client) {
			if( err ) {
				reject(err);
			} else {
				resolve( client );
			}
		});
	});
};

test.test2 = function() {
	logger.trace( "test2 : checkpasswd valid" );
	return new promise( function( resolve, reject) {
		soap.createClient("http://"+serverIP+'/wsdl', function(err, client) {
			var request = { CheckPasswordRequest:{ Authentifier: '03thomas.jabouley@viveris.fr', OrganizationalUnit:"User", Password:"test" }};
			client.CheckPassword(request, function(err, result, raw) {
				if( err ) {
					reject(err);
				} else {
					console.log(raw);
					resolve( result );
				}
		 	});
		});
	});
};

test.test3 = function() {
	logger.trace( "test3 : checkpasswd not valid" );
	return new promise( function( resolve, reject) {
		soap.createClient("http://"+serverIP+'/wsdl', function(err, client) {
			client.CheckPassword({ CheckPasswordRequest:{ OrganizationalUnit:"User", Password:"test" }}, function(err, result) {
				if( err ) {
					reject(err);
				} else {
					resolve( result );
				}
		 	});
		});
	});
};

test.test4 = function() {
	logger.trace( "test4 : checkpasswd not valid ( bad password )" );
	return new promise( function( resolve, reject) {
		soap.createClient("http://"+serverIP+'/wsdl', function(err, client) {
			client.CheckPassword( { CheckPasswordRequest:{ Authentifier: 'thomas.jabouley@viveris.fr', OrganizationalUnit:"User", Password:"test2" }} , function(err, result) {
				if( err ) {
					reject(err);
				} else {
					resolve( result );
				}
			});
		});
	});
};

test.test5 = function() {
	logger.trace( "test2 : checkpasswd valid" );
	return new promise( function( resolve, reject) {
		soap.createClient("http://"+serverIP+'/wsdl', function(err, client) {
			client.CheckPassword({ Authentifier: '03thomas.jabouley@viveris.fr', OrganizationalUnit:"User", Password:"test" }, function(err, result) {
				if( err ) {
					reject(err);
				} else {
					resolve( result );
				}
			});
		});
	});
};

var xml = require('fs').readFileSync('checkpassword.wsdl', 'utf8');
var server = http.createServer(function(req,res) {
	logger.trace("request to http server", req.method, req.url );
	var fn = test[req.url.slice(1)];
	var done = false;
	if( req.method === "GET" && fn ) {
		done = true;
		fn()
			.then( function( result ) {
				res.end( JSON.stringify(result,"",4));
			})
			.catch( function(err) {
				if( err.body ) {
					// var serverName =  req.headers.host === "localhost"?"localhost:1234":req.headers.host;
					var serverName = req.headers.host;
					var errXML = swig.render( err.body, { locals: { server: serverName } } );
					res.writeHead(200, {
						'Content-Length': errXML.length,
						'Content-Type': 'application/xml' });
					res.end( errXML );
				} else {
					res.end("err \n" + JSON.stringify(result, "", 4));
				}
			});
	}
	
	if( req.method === "GET" && ["/checkpassword.wsdl","/wsdl"].indexOf(req.url) !== -1 ) {
		logger.trace("get the wsdl file");
		done = true;
		logger.debug("host",req.headers.host );
		// var serverName =  req.headers.host === "localhost"?"localhost:1234":req.headers.host;
		var serverName = req.headers.host;
		var wsdl = swig.render( xml, { locals: { server: serverName } } );
		res.writeHead(200, {
			'Content-Length': wsdl.length,
			'Content-Type': 'application/xml' });
		res.end( wsdl );
	}
	if( !done ) {
		res.statusCode = 404;
		res.end("404: Not Found: "+req.url);
	}
});

server.listen(8010, function() {
	var pkg = require("./package.json");
	logger.info("-----------------------------------------------");
	logger.info("server wsdl v"+pkg.version+" listen on 8010" );
	logger.info("-----------------------------------------------");
	var wsdl = swig.render( xml, { locals: { server: serverIP } } );
	soap.listen(server, '/login', CheckPassword, wsdl);
});



SOAPservice.js
var http = require("http"),
	program = require("commander"),
	soap = require("soap"),
	RSVP = require("rsvp"),
	promise = require("rsvp").Promise,
	Logger = require("logger"),
	request = require("request"),
	Etcd = require("node-etcd"),
	swig = require('swig');

var pkg = require('./package.json');
var logger = new Logger("WSDL service");
var config = {};
var HHRPros = {};   // etcd HHRPro registred instances
var loop = setInterval(getHHRServices, 5000);

program
	.version(pkg.version)
	.usage('[options] [dir]')
	.option('-h, --host [value]', 'etcd server [172.17.42.1]', '172.17.42.1')
	.option('-c, --config <config>', 'configuration file', String)
	.parse(process.argv);

if (program.config) {
	if (!program.config.match(/^\//)) {
		program.config = path.join(__dirname + '/..', program.config);
	}
	var tmp = require(program.config);
	if (!tmp.serverIP) {
		logger.error("no serverIP defined");
		process.exit(1);
	} else {
		config.serverIP = tmp.serverIP;
	}
} else {
	logger.error("you must provide a config file");
	process.exit(1);
}

// the etcd client
var etcd = new Etcd(program.host, '4001');

function getHHRServices() {
	etcd.get('/services', {recursive: true}, function (err, resp) {
		if (err) {
			console.error("error ", err);
		} else {
			clearInterval(loop);
			if (resp.node.nodes && resp.node.nodes.length ) {
				resp.node.nodes.forEach(function (node) {
					if (node.key === "/services/HHR-Pro" && node.nodes) {
						HHRPros[node.nodes[0].key] = JSON.parse(node.nodes[0].value);
					}
				});
			}

			watcher = etcd.watcher("/services", null, {recursive: true});
			watcher.on("change", function (action) {
				switch (action.action) {
					case 'set':
						if (action.node.key.match(/^\/services\/HHR-Pro/)) {
							HHRPros[action.node.key] = JSON.parse(action.node.value);
						}
						break;
					case 'delete':
						if (action.node.key.match(/^\/services\/HHR-Pro/)) {
							delete HHRPros[action.node.key];
						}
						break;
				}
			});
		}
	});
}

var CheckPassword = {
	loginservice: {
		loginserviceSOAP: {
			CheckPassword: function (CheckPasswordRequest, cb) {
				var soapReq = CheckPasswordRequest.CheckPasswordRequest;

				if (!soapReq || !soapReq.Authentifier || !soapReq.OrganizationalUnit || !soapReq.Password) {
					logger.error("Requete d'identification invalide");
					throw {
						Fault: {
							Code  : {
								Value  : "soap:Sender",
								Subcode: {value: "rpc:BadArguments"}
							},
							Reason: {Text: "Requete invalide"}
						}
					};
				}

				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};

				function testPasswd(HHRPro, Authentifier, Password) {
					return new promise(function (resolve, reject) {
						request({
							url    : "http://" + HHRPro.dns + "/api/checkpasswd",
							method : "POST",
							headers: {"content-type": "text/plain"},
							body   : JSON.stringify({login: Authentifier, password: Password})
						}, function (err, resp, body) {
							if (err) {
								logger.debug("resp checkpasswd err", err);
								resolve({valid: false});
							} else {
								try {
									var isvalid = JSON.parse(body);
									resolve(isvalid);
								} catch (err) {
									resolve({valid: false});
								}
							}
						});
					});
				}

				var instances = [];
				for (var HHRPro in HHRPros) {
					instances.push(HHRPros[HHRPro]);
				}
				var promises = instances.map(function (HHRPro) {
					return testPasswd(HHRPro, soapReq.Authentifier, soapReq.Password);
				});

				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};
				// @todo valid is a function that validate the user:

				RSVP.all(promises)
					.then(function (results) {
						var valid = false;
						for (var i = 0; i < results.length; i++) {
							valid = valid || results[i].valid;
						}
						if (valid) {
							logger.info("Authentifier valid", soapReq);
							CheckPasswordOut.IsValid = true;
							CheckPasswordOut.CheckPasswordUserInfo = "Authentifier=" + soapReq.Authentifier + ";OrganizationalUnit=" + soapReq.OrganizationalUnit + ";";
						} else {
							logger.alert("Authentifier not valid", soapReq);
							CheckPasswordOut.IsValid = false;
						}
						logger.trace({CheckPasswordResponse: CheckPasswordOut});
						cb({CheckPasswordResponse: CheckPasswordOut});
					});
			}
		}
	}
};

var xml = require('fs').readFileSync(__dirname + '/checkpassword.wsdl', 'utf8');
var server = http.createServer(function (req, res) {
	logger.trace("request to http server", req.method, req.url);

	var done = false;

	if (req.method === "GET" && ["/checkpassword.wsdl", "/wsdl"].indexOf(req.url) !== -1) {
		logger.trace("get the wsdl file");
		done = true;
		logger.debug("host", req.headers.host);
		// var serverName =  req.headers.host === "localhost"?"localhost:1234":req.headers.host;
		var serverName = req.headers.host;
		var wsdl = swig.render(xml, {locals: {server: serverName}});
		res.writeHead(200, {
			'Content-Length': wsdl.length,
			'Content-Type'  : 'application/xml'
		});
		res.end(wsdl);
	}
	if (!done) {
		res.statusCode = 404;
		res.end("404: Not Found: " + req.url);
	}
});

server.listen(8010, function () {
	logger.info("-----------------------------------------------");
	logger.info("server wsdl v" + pkg.version + " listen on 8010");
	logger.info("-----------------------------------------------");
	var wsdl = swig.render(xml, {locals: {server: config.serverIP}});
	soap.listen(server, '/login', CheckPassword, wsdl);
	// Watch for HHRPro instances
	getHHRServices();
});



SOAPservice-prod.js
var http = require("http"),
	program = require("commander"),
	soap = require("soap"),
	RSVP = require("rsvp"),
	promise = require("rsvp").Promise,
	Logger = require("logger"),
	request = require("request"),
	Etcd = require("node-etcd"),
	swig  = require('swig');

var pkg     = require('./package.json');
var logger = new Logger("WSDL service");
var config = {};
var HHRPros = {};   // etcd HHRPro registred instances
var loop = setInterval( getHHRServices, 5000);

program
	.version(pkg.version)
	.usage('[options] [dir]')
	.option('-h, --host [value]','etcd server [172.17.42.1]','172.17.42.1')
	.option('-c, --config <config>', 'configuration file', String )
	.parse(process.argv);

if( program.config ) {
	if( !program.config.match(/^\//) ) {
		program.config = path.join(__dirname+'/..',program.config );
	}
	var tmp = require(program.config);
	if( !tmp.serverIP ) {
		logger.error("no serverIP defined");
	 	process.exit(1);
	} else {
		config.serverIP = tmp.serverIP;
	}
} else {
	logger.error("you must provide a config file");
	process.exit(1);
}

// the etcd client
var etcd = new Etcd( program.host, '4001' );

function getHHRServices() {
	etcd.get('/services', { recursive: true }, function( err, resp ) {
		if(err) {
		 	console.error("error ",err);
		} else {
			clearInterval(loop);
			if( resp.node.nodes ) {
				if(resp.node.nodes.length) {
					resp.node.nodes.forEach( function(node) {
						if( node.key === "/services/HHR-Pro" && node.nodes ) {
							HHRPros[node.nodes[0].key] = JSON.parse(node.nodes[0].value);
						}
					});
				}
			}
			
			watcher = etcd.watcher("/services", null, { recursive:true} );
			watcher.on("change", function( action ) {
				switch( action.action ) {
					case 'set':
						if( action.node.key === "/services/HHR-Pro") {
							HHRPros[action.node.nodes[0].key] = JSON.parse(action.node.nodes[0].value);
						}
						break;
					case 'delete':
						if( action.node.key === "/services/HHR-Pro") {
							delete HHRPros[action.node.nodes[0].key];
						}
						break;
				}
			});
		}
	});
}

var CheckPassword = {
	loginservice: {
		loginserviceSOAP: {
			CheckPassword: function (CheckPasswordRequest, cb) {
				var soapReq = CheckPasswordRequest.CheckPasswordRequest;
				
				if (!soapReq || !soapReq.Authentifier || !soapReq.OrganizationalUnit || !soapReq.Password) {
					logger.error("Requete d'identification invalide");
					throw {
						Fault: {
							Code  : {
								Value  : "soap:Sender",
								Subcode: {value: "rpc:BadArguments"}
							},
							Reason: {Text: "Requete invalide"}
						}
					};
				}
				
				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};
				
				function testPasswd( HHRPro, Authentifier, Password ) {
					return new promise( function(resolve, reject) {
						request({
							url    : "http://"+HHRPro.dns+"/api/checkpasswd",
							method : "POST",
							headers: {"content-type": "text/plain"},
							body   : JSON.stringify({login: Authentifier, password: Password})
						}, function (err, resp, body) {
							if (err) {
								logger.debug("resp checkpasswd err", err);
								resolve( { valid: false });
							} else {
								try {
									var isvalid = JSON.parse(body);
									resolve( isvalid );
								} catch( err ) {
									resolve( { valid: false });
								}
							}
						});
					});
				}
				
				var instances = [];
				for( var HHRPro in HHRPros ) {
					instances.push( HHRPros[HHRPro] );
				}
				var promises = instances.map(function(HHRPro){
					return testPasswd(HHRPro, soapReq.Authentifier, soapReq.Password );
				});
				
				var CheckPasswordOut = {IsValid: false, CheckPasswordUserInfo: ""};
				// @todo valid is a function that validate the user:
								
				RSVP.all( promises )
					.then( function(results) {
						var valid = false;
						for( var i=0; i<results.length; i++ ) {
							valid = valid || results[i].valid;
						}
						if (valid) {
							logger.info("Authentifier valid", soapReq);
							CheckPasswordOut.IsValid = true;
							CheckPasswordOut.CheckPasswordUserInfo = "Authentifier=" + soapReq.Authentifier + ";OrganizationalUnit="
								 + soapReq.OrganizationalUnit + ";";
						} else {
							logger.alert("Authentifier not valid", soapReq);
							CheckPasswordOut.IsValid = false;
						}
						logger.trace( { CheckPasswordResponse: CheckPasswordOut } );
						cb({ CheckPasswordResponse: CheckPasswordOut });
					});
			}
		}
	}
};

var xml = require('fs').readFileSync(__dirname + '/checkpassword.wsdl', 'utf8');
var server = http.createServer(function(req,res) {
	logger.trace("request to http server", req.method, req.url );
	
	var done = false;
	
	if( req.method === "GET" && ["/checkpassword.wsdl","/wsdl"].indexOf(req.url) !== -1 ) {
		logger.trace("get the wsdl file");
		done = true;
		logger.debug("host",req.headers.host );
		// var serverName =  req.headers.host === "localhost"?"localhost:1234":req.headers.host;
		var serverName = req.headers.host;
		var wsdl = swig.render( xml, { locals: { server: serverName } } );
		res.writeHead(200, {
			'Content-Length': wsdl.length,
			'Content-Type': 'application/xml' });
		res.end( wsdl );
	}
	if( !done ) {
	        res.statusCode = 404;
	        res.end("404: Not Found: "+req.url);
	}
});

server.listen(8010, function() {
	logger.info("-----------------------------------------------");
	logger.info("server wsdl v"+pkg.version+" listen on 8010" );
	logger.info("-----------------------------------------------");
	var wsdl = swig.render( xml, { locals: { server: config.serverIP } } );
	soap.listen(server, '/login', CheckPassword, wsdl);
	// Watch for HHRPro instances
	getHHRServices();
});